### PR TITLE
Storm-POMDP update to latest version

### PR DESCRIPTION
**Pull request by AlexBork**
This PR contains the updates to Storm-POMDP with regards to the quantitative analysis. 
The main functionality added:
- Improvements to cut-offs for the under-approximation
- Inclusion of the clipping approach as described in the TACAS 2022 paper
- Extension of the implementation to interact with PAYNT for the SAYNT approach (to appear at CAV 2023)
- Streamlining of the AR-loop implementation: the one-shot computation is now fully integrated

Furthermore the PR brings large parts of Storm-POMDP to the latest standards used in the rest of Storm.

Thanks to @SvStein for helping out with a lot of the refactoring!

### CODE REVIEW

**Comment by AlexBork** [comment]
My suggestion is to add Storm-POMDP to the formatting check and run the auto-formatting after the merge. That should make the review easier.

File: README.md [hunk 1]
```diff
@@ -1,10 +1,9 @@
 Storm - A Modern Probabilistic Model Checker
 ============================================
 
-[![Build Status](https://github.com/moves-rwth/storm/workflows/Build%20Test/badge.svg)](https://github.com/moves-rwth/storm/actions)
-[![GitHub release](https://img.shields.io/github/release/moves-rwth/storm.svg)](https://github.com/moves-rwth/storm/releases/)
-[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1181896.svg)](https://doi.org/10.5281/zenodo.1181896)
-
+<!-- [![Build Status](https://github.com/moves-rwth/storm/workflows/Build%20Test/badge.svg)](https://github.com/moves-rwth/storm/actions) -->
```
Comment (volkm) [inline README.md:4]: This should be reverted.

File: CMakeLists.txt [hunk 2]
```diff
@@ -581,4 +581,4 @@ install(FILES ${CMAKE_BINARY_DIR}/stormConfig.install.cmake DESTINATION ${CMAKE_
 install(FILES ${CMAKE_BINARY_DIR}/stormConfigVersion.cmake DESTINATION ${CMAKE_INSTALL_DIR})
 install(EXPORT storm_Targets FILE stormTargets.cmake DESTINATION ${CMAKE_INSTALL_DIR})
 
-include(StormCPackConfig.cmake)
\ No newline at end of file
+include(StormCPackConfig.cmake)
```
Comment (volkm) [inline CMakeLists.txt:584]: Can we revert this?

File: src/storm/modelchecker/prctl/helper/SparseMdpPrctlHelper.cpp [hunk 3]
```diff
@@ -379,7 +379,9 @@ SparseMdpHintType<ValueType> computeHints(Environment const& env, SolutionType c
     if (!result.eliminateEndComponents) {
         extractValueAndSchedulerHint(result, transitionMatrix, backwardTransitions, maybeStates, selectedChoices, hint, result.uniqueSolution);
     } else {
-        STORM_LOG_WARN_COND(hint.isEmpty(), "A non-empty hint was provided, but its information will be disregarded.");
+        if (hint.isEmpty()) {
+            STORM_LOG_TRACE("Warn A non-empty hint was provided, but its information will be disregarded.");
+        }
     }
```
Comment (volkm) [inline src/storm/modelchecker/prctl/helper/SparseMdpPrctlHelper.cpp:385]: What is the reason for this change? Too much output otherwise?

Comment (AlexBork) [reply src/storm/modelchecker/prctl/helper/SparseMdpPrctlHelper.cpp:385]: Exactly, in particular because it is not really something a user should be concerned about.

File: README.md (hunk 1 above)
Comment (AlexBork) [reply README.md]: Done

File: CMakeLists.txt (hunk 2 above)
Comment (AlexBork) [reply CMakeLists.txt]: Done.

**Comment by AlexBork** [comment]
Yes, it does. I originally applied the auto-format, but got confused why it didn't apply to the POMDP code so I reverted. I can apply formatting now or we can merge and then also apply formatting to the POMDP code, depending on whether it is acceptable to have the formatting check fail in the meantime.

File: src/storm-pomdp/modelchecker/PreprocessingPomdpValueBoundsModelChecker.cpp [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+    namespace pomdp {
+        namespace modelchecker {
+            template <typename ValueType>
+            PreprocessingPomdpValueBoundsModelChecker<ValueType>::PreprocessingPomdpValueBoundsModelChecker(storm::models::sparse::Pomdp<ValueType> const& pomdp, storm::solver::MinMaxMethod minMaxMethod) : pomdp(pomdp) {
+                // Setup checker environment
+                mcEnvironment = storm::Environment();
+                mcEnvironment.solver().setLinearEquationSolverType(storm::solver::EquationSolverType::Topological, true);
+                mcEnvironment.solver().minMax().setMethod(minMaxMethod);
+                mcEnvironment.solver().minMax().setPrecision(storm::utility::convertNumber<storm::RationalNumber>(1e-9));
+                mcEnvironment.solver().topological().setUnderlyingEquationSolverType(storm::solver::EquationSolverType::Eigen);
```
Comment (tquatmann) [inline src/storm-pomdp/modelchecker/PreprocessingPomdpValueBoundsModelChecker.cpp:27]: Why do we require topological + Eigen for DTMCs and (non-topological) SVI for MDPs?
Maybe, what we actually want here is something like `setForceExact` or `setForceSoundness` in `SolverEnvironment` ?
In any case, it would be nice to elaborate a bit on this.

Usually, we pass a `storm::Environment cons& env` to all the methods that need an environment, i.e., it should be added as the first argument of `getValueBounds` and `getExtremeValueBound`.  Then, the calling method needs to take care of configuring the environment which makes things a bit more transparent.

There might be reasons to do it differently here that I don't see right now, though.

File: src/storm-pomdp/modelchecker/BeliefExplorationPomdpModelCheckerOptions.h [hunk 5]
```diff
@@ -36,9 +46,18 @@ namespace storm {
...
+                uint64_t clippingGridRes = 2;
+
+                bool skipHeuristicSchedulers = false;
+
                 ValueType numericPrecision = storm::NumberTraits<ValueType>::IsExact ? storm::utility::zero<ValueType>() : storm::utility::convertNumber<ValueType>(1e-9); /// Used to decide whether two beliefs are equal
                 bool dynamicTriangulation = true; // Sets whether the triangulation is done in a dynamic way (yielding more precise triangulations)
+
+                storm::builder::ExplorationHeuristic explorationHeuristic = storm::builder::ExplorationHeuristic::BreadthFirst;
+
+                storm::solver::MinMaxMethod preProcMinMaxMethod = storm::solver::MinMaxMethod::SoundValueIteration;
```
Comment (tquatmann) [inline src/storm-pomdp/modelchecker/BeliefExplorationPomdpModelCheckerOptions.h:59]: I suggest to have a `storm::Environment preProcEnvironment` instead if this really needs to be configured "from the outside"

File: src/storm-pars/transformer/SparseParametricDtmcSimplifier.cpp [hunk 6]
```diff
@@ -34,8 +34,9 @@ namespace storm {
             std::pair<storm::storage::BitVector, storm::storage::BitVector> statesWithProbability01 = storm::utility::graph::performProb01(this->originalModel, phiStates, psiStates);
             // Only consider the maybestates that are reachable from one initial state without hopping over a target (i.e., prob1) state
             storm::storage::BitVector reachableGreater0States = storm::utility::graph::getReachableStates(this->originalModel.getTransitionMatrix(), this->originalModel.getInitialStates() & ~statesWithProbability01.first, ~statesWithProbability01.first, statesWithProbability01.second);
-            storm::storage::BitVector maybeStates = reachableGreater0States & ~statesWithProbability01.second;
-            
+            //storm::storage::BitVector maybeStates = reachableGreater0States & ~statesWithProbability01.second;
+            storm::storage::BitVector maybeStates = ~(statesWithProbability01.second | statesWithProbability01.first);
```
[the remark this answers is not in the document]
Comment (AlexBork) [reply src/storm-pars/transformer/SparseParametricDtmcSimplifier.cpp:38]: Yes, I think that was the point. However, the native parametric transformation checking is not supposed to be part of the release, so I'll revert this change for now. I'll also keep your comment in mind for later implementations.

**Review by tquatmann — COMMENTED** [review]
Besides my two minor comments, I approve 👍 

[I forward declared the environment to optimise compile times a little.]

**Comment by AlexBork** [comment]
I incorporated Tim's comments and fixed a small bug triggered by the new VI implementation in conjunction with infinite reward values. The current state of the main branch is also merged.

**Review by tquatmann — APPROVED** [review]
